web/src/stream_data.ts

```diff
@@ -706,7 +706,35 @@ export function get_current_user_and_their_bots_with_post_messages_permission(
 }
 
 export function can_access_stream_email(sub: StreamSubscription): boolean {
-    return get_current_user_and_their_bots_with_post_messages_permission(sub).length > 0;
+    // Web public streams are accessible to everyone
+    if (sub.is_web_public) {
+        return true;
+    }
+
+    // Subscribed users (including guests) have access to stream email
+    if (sub.subscribed) {
+        return true;
+    }
+
+    // For non-subscribed users, check if they have post message permissions
+    // Users without post permissions should not have email access
+    if (get_current_user_and_their_bots_with_post_messages_permission(sub).length === 0) {
```

Comment:
This condition should be at the top. Please see `access_stream_for_send_message` code in `zerver/lib/streams.py` as a reference for how the permissions should be checked.